### PR TITLE
fix IMA bug in FA2

### LINKED ISSUE #173953: [CUDA][Flash Attention][SDPA] Flash Attention 2.8.3 submodule upgrade introduces IMAs on B200
### 🐛 Describe the bug

After a Flash Attention submodule upgrade we started seeing IMAs on a stable diffusion regression test workload.
B200 seems to show this, and A100 as well but I have not attempted to reproduce. The same issue is suspected on both.

Bisected to this offending commit, https://github.com/pytorch/pytorch/pull/170703
The next submodule upgrade (https://github.com/pytorch/pytorch/pull/173214) also does not appear to fix it

Run e.g., `PYTORCH_NO_CUDA_MEMORY_CACHING=1 compute-sanitizer python3 repro.py` to reproduce, does not repro if the FA 2.8.3 upgrade is rolled back. @Aidyn-A showed that the removal of head-dim 160 kernels seems to be responsible: https://github.com/Dao-AILab/flash-attention/commit/ce2127207f5764d79e6a06751f20532cca1a4720 
as `void pytorch_flash::flash_bwd_dot_do_o_kernel<true, Flash_bwd_kernel_traits<160, 64, 64, 8, 4, 4, 4, false, false, cutlass::half_t, Flash_kernel_traits<160, 64, 64, 8, cutlass::half_t> > >(pytorch_flash::Flash_bwd_params)
` did not have the IMA while 
`void pytorch_flash::flash_bwd_dot_do_o_kernel<true, Flash_bwd_kernel_traits<192, 64, 64, 8, 4, 2, 2, false, false, cutlass::half_t, Flash_kernel_traits<192, 64, 64, 8, cutlass::half_t> > >(pytorch_flash::Flash_bwd_params)
` does.

Repro generated by Claude code:
```Python
#!/usr/bin/env python3
"""
Minimal reproducer for Flash Attention illegal memory access (IMA) issue.

This script reproduces the IMA seen in Stable Diffusion training with the exact
tensor shapes used in the model. The IMA appears to occur in the backward pass,
specifically in flash_bwd_dot_do_o_kernel.

Based on Stable Diffusion v1.5 configuration:
- attention_head_dim: 8
- cross_attention_dim: 768
- Resolution: 512x512 -> latent space 64x64 -> sequence length 4096
- Batch size: 16
- Mixed precision: fp16
"""

import torch
import torch.nn.functional as F

def test_flash_attention_backward(
    batch_size=16,
    num_heads=8,
    seq_len=4096,  # 64x64 for latent resolution
    head_dim=40,  # 320 // 8 = 40 for first attention block
    dtype=torch.float16,
    device="cuda"
):
    """
    Test Flash Attention backward pass with shapes from SD v1.5.

    This uses the self-attention configuration from the first down block
    where channels=320, num_heads=8, head_dim=40.
    """
    print(f"Testing Flash Attention backward with:")
    print(f"  batch_size={batch_size}, num_heads={num_heads}")
    print(f"  seq_len={seq_len}, head_dim={head_dim}")
    print(f"  dtype={dtype}, device={device}")

    # Create input tensors
    query = torch.randn(batch_size, num_heads, seq_len, head_dim,
                        dtype=dtype, device=device, requires_grad=True)
    key = torch.randn(batch_size, num_heads, seq_len, head_dim,
                      dtype=dtype, device=device, requires_grad=True)
    value = torch.randn(batch_size, num_heads, seq_len, head_dim,
                        dtype=dtype, device=device, requires_grad=True)

    print(f"  query shape: {query.shape}")
    print(f"  key shape: {key.shape}")
    print(f"  value shape: {value.shape}")

    # Force Flash Attention backend
    from torch.nn.attention import SDPBackend, sdpa_kernel

    print("\nRunning forward pass with Flash Attention...")
    with sdpa_kernel([SDPBackend.FLASH_ATTENTION]):
        output = F.scaled_dot_product_attention(
            query, key, value,
            attn_mask=None,
            dropout_p=0.0,
            is_causal=False
        )

    print(f"  output shape: {output.shape}")
    print("Forward pass completed successfully!")

    # Backward pass - this is where IMA typically occurs
    print("\nRunning backward pass (this may trigger IMA)...")
    grad_output = torch.randn_like(output)
    output.backward(grad_output)

    print("Backward pass completed successfully!")
    print(f"  query.grad shape: {query.grad.shape}")
    print(f"  key.grad shape: {key.grad.shape}")
    print(f"  value.grad shape: {value.grad.shape}")


def test_flash_attention_cross_attention_backward(
    batch_size=16,
    num_heads=8,
    seq_len=4096,  # 64x64 for latent resolution
    context_len=77,  # CLIP max sequence length
    head_dim=40,  # 320 // 8 = 40
    dtype=torch.float16,
    device="cuda"
):
    """
    Test Flash Attention backward pass with cross-attention shapes.

    This tests cross-attention where query comes from latent space
    and key/value come from text encoder (CLIP).
    """
    print(f"\nTesting Flash Attention cross-attention backward with:")
    print(f"  batch_size={batch_size}, num_heads={num_heads}")
    print(f"  seq_len={seq_len}, context_len={context_len}, head_dim={head_dim}")
    print(f"  dtype={dtype}, device={device}")

    # Query from latent space, K/V from text encoder
    query = torch.randn(batch_size, num_heads, seq_len, head_dim,
                        dtype=dtype, device=device, requires_grad=True)
    key = torch.randn(batch_size, num_heads, context_len, head_dim,
                      dtype=dtype, device=device, requires_grad=True)
    value = torch.randn(batch_size, num_heads, context_len, head_dim,
                        dtype=dtype, device=device, requires_grad=True)

    print(f"  query shape: {query.shape}")
    print(f"  key shape: {key.shape}")
    print(f"  value shape: {value.shape}")

    # Force Flash Attention backend
    from torch.nn.attention import SDPBackend, sdpa_kernel

    print("\nRunning forward pass with Flash Attention...")
    with sdpa_kernel([SDPBackend.FLASH_ATTENTION]):
        output = F.scaled_dot_product_attention(
            query, key, value,
            attn_mask=None,
            dropout_p=0.0,
            is_causal=False
        )

    print(f"  output shape: {output.shape}")
    print("Forward pass completed successfully!")

    # Backward pass
    print("\nRunning backward pass (this may trigger IMA)...")
    grad_output = torch.randn_like(output)
    output.backward(grad_output)

    print("Backward pass completed successfully!")
    print(f"  query.grad shape: {query.grad.shape}")
    print(f"  key.grad shape: {key.grad.shape}")
    print(f"  value.grad shape: {value.grad.shape}")


if __name__ == "__main__":
    print("=" * 80)
    print("Flash Attention Backward IMA Minimal Reproducer")
    print("=" * 80)

    if not torch.cuda.is_available():
        print("ERROR: CUDA not available!")
        exit(1)

    print(f"\nPyTorch version: {torch.__version__}")
    print(f"CUDA version: {torch.version.cuda}")
    print(f"Device: {torch.cuda.get_device_name(0)}")

    # Test 1: Self-attention with first block dimensions (320 channels)
    print("\n" + "=" * 80)
    print("TEST 1: Self-Attention (320 channels, head_dim=40)")
    print("=" * 80)
    try:
        test_flash_attention_backward(
            batch_size=16,
            num_heads=8,
            seq_len=4096,  # 64x64
            head_dim=40,   # 320 // 8
            dtype=torch.float16
        )
    except Exception as e:
        print(f"\n!!! ERROR IN TEST 1: {e}")
        import traceback
        traceback.print_exc()

    # Test 2: Self-attention with second block dimensions (640 channels)
    print("\n" + "=" * 80)
    print("TEST 2: Self-Attention (640 channels, head_dim=80)")
    print("=" * 80)
    try:
        test_flash_attention_backward(
            batch_size=16,
            num_heads=8,
            seq_len=1024,  # 32x32
            head_dim=80,   # 640 // 8
            dtype=torch.float16
        )
    except Exception as e:
        print(f"\n!!! ERROR IN TEST 2: {e}")
        import traceback
        traceback.print_exc()

    # Test 3: Self-attention with third block dimensions (1280 channels)
    print("\n" + "=" * 80)
    print("TEST 3: Self-Attention (1280 channels, head_dim=160)")
    print("=" * 80)
    try:
        test_flash_attention_backward(
            batch_size=16,
            num_heads=8,
            seq_len=256,   # 16x16
            head_dim=160,  # 1280 // 8
            dtype=torch.float16
        )
    except Exception as e:
        print(f"\n!!! ERROR IN TEST 3: {e}")
        import traceback
        traceback.print_exc()

    # Test 4: Cross-attention (320 channels)
    print("\n" + "=" * 80)
    print("TEST 4: Cross-Attention (320 channels, context_len=77)")
    print("=" * 80)
    try:
        test_flash_attention_cross_attention_backward(
            batch_size=16,
            num_heads=8,
            seq_len=4096,  # 64x64
            context_len=77,
            head_dim=40,   # 320 // 8
            dtype=torch.float16
        )
    except Exception as e:
        print(f"\n!!! ERROR IN TEST 4: {e}")
        import traceback
        traceback.print_exc()

    # Test 5: Cross-attention (640 channels)
    print("\n" + "=" * 80)
    print("TEST 5: Cross-Attention (640 channels, context_len=77)")
    print("=" * 80)
    try:
        test_flash_attention_cross_attention_backward(
            batch_size=16,
            num_heads=8,
            seq_len=1024,  # 32x32
            context_len=77,
            head_dim=80,   # 640 // 8
            dtype=torch.float16
        )
    except Exception as e:
        print(f"\n!!! ERROR IN TEST 5: {e}")
        import traceback
        traceback.print_exc()

    # Test 6: Cross-attention (1280 channels)
    print("\n" + "=" * 80)
    print("TEST 6: Cross-Attention (1280 channels, context_len=77)")
    print("=" * 80)
    try:
        test_flash_attention_cross_attention_backward(
            batch_size=16,
            num_heads=8,
            seq_len=256,   # 16x16
            context_len=77,
            head_dim=160,  # 1280 // 8
            dtype=torch.float16
        )
    except Exception as e:
        print(f"\n!!! ERROR IN TEST 6: {e}")
        import traceback
        traceback.print_exc()

    print("\n" + "=" * 80)
    print("All tests completed!")
    print("=" * 80)

```

### Versions

current source build

cc @ezyang @gchanan @kadeng @msaroufim @ptrblck @jerryzh168 @tinglvv @nWEIdia

### PR DESCRIPTION
**Pull request by liangel-02**
Fixes #173953 

A recent submodule upgrade of flash attention removed `flash_bwd_dot_do_o_kernel` for head dim 160 so now anything between 129-160 gets rounded up to the 192 kernel. The related change in the upstream flash repo is this commit: https://github.com/Dao-AILab/flash-attention/commit/ce2127207f5764d79e6a06751f20532cca1a4720

However, the current logic in `flash_api.cpp` still rounds the kernels by multiples of 32, so for hdim 160, the `dq_accum` buffer is still being allocated with 160 positions. But since the 192 kernel is being called, it ends up writing to garbage addresses. 

The solution is to update `flash_api.cpp` to match the upstream logic which now rounds with multiples of 64. 

Testing

Added more cases to `test_transformers.py`
Repro provided in original issue passes

### CODE REVIEW

**Review by drisspg — APPROVED** [review]
looks good I think that this might blow up tests alot, so maybe be a lil more targeted on which head dims you add

Or add a more targeted one that does not parametrize as much

**Comment by liangel-02** [comment]
cc @eqy

**Review by eqy — APPROVED** [review]

**Review by malfet — APPROVED** [review]
LGTM, but I though that adding more shapes doesn't really help, unless test is run with `compute-sanitizer`, isn't it?
